Riak - Create Bucket type effector

bostko opened the pull request:
This is the initial pull request for supporting riak bucket types in brooklyn.
http://docs.basho.com/riak/latest/dev/advanced/bucket-types/

Bucket types are more used for dynamic configurations, for example changing easily properties of a running cluster.
Riak supports adding, updating and deleting bucket types.

I added an effector which creates a bucket type.
I still can not figure out how can we add some monitoring /sensors/ to those bucket types?

aledsage commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java:
```diff
@@ -183,18 +183,21 @@
+    void commitCluster();
 
+    @Effector(description = "Apply bucket type")
+    void applyBucketType(@EffectorParam(name = "bucketTypeName") String bucketTypeName,
```
Is it feasible to have an integration and/or live test that uses this (including subsequently using the bucket-type that is created + activated)?

aledsage commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNodeSshDriver.java:
```diff
@@ -474,6 +475,18 @@ public void commitCluster() {
     }
 
     @Override
+    public void applyBucketType(String bucketTypeName, String bucketTypeProperties) {
```
perhaps "createBucketType" would be a better method name, given that we all Riak's `bucket-type create ...`? (even though it means create or modify - still arguably better to be consistent with Riak's commands and documentation). Same goes for effector method on RiakNode as well (but the effector description can say "Create or modify a bucket type before activation" [1]).

Do we also need to activate the bucket type: "Like all bucket types, this type needs to be activated to be usable within the cluster." [1]. That is, call `riak-admin bucket-type activate ...`?

[1] http://docs.basho.com/riak/latest/dev/advanced/bucket-types/

aledsage commented:
@bostko not sure what has happened to this PR - github thinks it includes the two commits that have been merged to master in other pull requests! Out of interest, how did you `git pull` or `git rebase` your branch?

bostko commented:
Thank you for your comment.
It is still not ready for merging I just wanted some feedback

bostko commented:
I included also a fix for the cluster join command
http://docs.basho.com/riak/latest/ops/building/basic-cluster-setup/#Add-a-Second-Node-to-Your-Cluster

neykov commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNodeSshDriver.java:
```diff
@@ -395,12 +390,12 @@ public void joinCluster(String nodeName) {
                 ScriptHelper joinClusterScript = newScript("joinCluster")
                         .body.append(sudo(format("%s cluster join %s", getRiakAdminCmd(), nodeName)))
+                        .body.append(sudo(format("%s cluster plan", getRiakAdminCmd())))
+                        .body.append(sudo(format("%s cluster commit", getRiakAdminCmd())))
```
Looking at the driver API looks like .commitCluster() is meant to be called as a separate command. Why call commit here and not for the other methods?

bostko replied:
Very good note.
The point here is that this clusterCommit() method is made as effector.
If we go to the use case of that command it is as following: executing one of the following effectors: leaveCluster(), joinCluster(), removeNode() and maybe recoverFailedNode()
So let's say we want to support executing these commands and then invoking commitCluster() at the end.

However when setting up a new cluster blueprint it will lead that at the end of blueprint initialization the user should manually go to the UI and click on commitCluster()
The way we could workaround this is by introducing new version of the methods leaveCluster(), joinCluster(), removeNode() like leaveClusterEffector(), joinClusterEffector(), removeNodeEffector() which is without invoking commitCluster() and leave the current methods as they are.
That way deploying and operations queue would work but then it won't be very well supported in the UI, it will be easy for the user to miss to invoke commitCluster().

ahgittin replied:
agree, it feels like our default behaviour should be to commit.

we could (in time) add an optional parameter "commitChange" which defaults to true, that way if someone wanted to they could batch/preview changes.

i'd say add a TODO with this (and reference the commit which includes the commitCluster effector, in case we want to revive that we can copy-paste from here.)

bostko replied:
Actually I found why it is implemented that way.
I think the initial idea was to support cluster queue operations. It is correct not to commit just after the cluster join command. Other than that it was executed for one of the members.
Instead of that commit was executed after the joinings are done and then in the RiakCluster.start at the end it was executed clusterCommit()
The bug here is that the commit is not executed properly because every time joining was executed against random member of the ring instead of using one, so commiting should be done against the nodes or the node which triggered cluster join. This is the actual bug. And the other bug is may be that leaveCluster had plan and commit command which is unnecessary.

My point is that it could support queue cluster operations but the task requires bigger effort.

neykov commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNodeSshDriver.java:
```diff
@@ -474,6 +463,79 @@ public void commitCluster() {
+                        getRiakAdminCmd(),
+                        bucketTypeName,
+                        escapeLiteralForDoubleQuotedBash(bucketTypeProperties))))
+                .noExtraOutput()
```
no need for `noExtraOutput` if you are not parsing stdout

bostko replied:
You are right

neykov commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNodeSshDriver.java:
```diff
@@ -542,4 +602,10 @@ public String getOsMajorVersion() {
+
+    private void addRiakOnPath(ScriptHelper scriptHelper) {
+        Map<String, String> newPathVariable = ImmutableMap.of("PATH", sbinPath);
+//        log.warn("riak command not found on PATH. Altering future commands' environment variables from {} to {}", getShellEnvironment(), newPathVariable);
```
Why comment it out? If really not needed better remove it altogether.

bostko replied:
For distributions that require this, It is annoying to show it all the time in the log

neykov commented:
Brooklyn side changes looking good, can't comment on the riak commands.

ahgittin commented:
btw @bostko to answer your question in original post:  if we want sensors/effectors for individual bucket types what i would do is add entities for those bucket types (and same for buckets, and for databases in DBMSs etc).

not needed at this stage.

ahgittin commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java:
```diff
@@ -183,18 +183,31 @@
-    @Effector(description = "Commit changes made to the Riak cluster")
-    public void commitCluster();
+    @Effector(description = "Create bucket type")
+    void bucketTypeCreate(@EffectorParam(name = "bucketTypeName") String bucketTypeName,
```
verb first is the usual pattern for methods isn't it?   so `createBucketType`

bostko replied:
I am following the riak commands flow.
This way it will also stay sorted in the UI, so bucket stuff will be next to each other

ahgittin replied:
as discussed -- you are following the riak CLI syntax, with nice side effect that effectors are nicely sorted.  makes a lot of sense.  disregard these comments.  :)

ahgittin commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java:
```diff
@@ -183,18 +183,31 @@
 
-    @Effector(description = "Commit changes made to the Riak cluster")
-    public void commitCluster();
+    @Effector(description = "Create bucket type")
```
i hate descriptions and javadoc which simply restate the name.  say `Creates a new bucket type with the given name and properties, with properties in <?what kind of ?> format`.  (or should properties be a map?)

ahgittin commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java:
```diff
@@ -183,18 +183,31 @@
+                          @EffectorParam(name = "bucketTypeProperties") String bucketTypeProperties);
 
+    @Effector(description = "List bucket types")
+    List<String> bucketTypeList();
```
as above, call this `listBucketTypes` ?  (note the plural.)  and expand the description, at least a little.  `List the names all bucket types in the system.`

ahgittin commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java:
```diff
@@ -183,18 +183,31 @@
+    List<String> bucketTypeList();
+
+    @Effector(description = "Status bucket type")
+    List<String> bucketTypeStatus(@EffectorParam(name = "bucketTypeName") String bucketTypeName);
```
don't understand this -- is it `lookupBucketTypeStatus` ?  what is the meaning of the return `List<String>` ?

ahgittin commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java:
```diff
@@ -183,18 +183,31 @@
+    List<String> bucketTypeStatus(@EffectorParam(name = "bucketTypeName") String bucketTypeName);
+
+    @Effector(description = "Update bucket type")
+    void bucketTypeUpdate(@EffectorParam(name = "bucketTypeName") String bucketTypeName,
```
verb first again

ahgittin commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java:
```diff
@@ -183,18 +183,31 @@
+                          @EffectorParam(name = "bucketTypeProperties") String bucketTypeProperties);
+
+    @Effector(description = "Activate bucket type")
+    void bucketTypeActivate(@EffectorParam(name = "bucketTypeName") String bucketTypeName);
```
and again

ahgittin commented on software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNodeSshDriver.java:
```diff
@@ -445,32 +438,81 @@ public void removeNode(String nodeName) {
-                    .body.append(sudo(format("%s cluster plan", getRiakAdminCmd())))
-                    .body.append(sudo(format("%s cluster commit", getRiakAdminCmd())))
-                    .failOnNonZeroResultCode();
+    public void bucketTypeCreate(String bucketTypeName, String bucketTypeProperties) {
```
again, change method names here to match effector rename

ahgittin commented:
looks good.  add the TODO and expand descriptions, then i think good to merge.

can test whether we can abandon `sudo` in a subsequent PR.

bostko commented:
I fixed the bucket type effectors' description and I added a TODO in https://github.com/bostko/incubator-brooklyn/commit/343ed278035e4fe58d9273add9c4f16076d09a24

ahgittin commented:
LGTM.  Needs a lot of testing, but going ahead and merging.
